test/metabase/api/action_test.clj

```diff
@@ -633,3 +639,99 @@
           (testing "error if actions is disabled"
             (is (= "Actions are not enabled."
                  (:message (mt/user-http-request :crowberto :get 400 (format "action/%d/execute" delete-action-id) :parameters (json/encode {:id 1})))))))))))
+
+(mt/defdataset action-error-handling
+  [["group"
+    [{:field-name "name" :base-type :type/Text :not-null? true}
+     {:field-name "ranking" :base-type :type/Integer :not-null? true :unique? true}]
+    [["admin" 1]
+     ["user" 2]]]
+   ["user"
+    [{:field-name "name" :base-type :type/Text :not-null? true}
+     {:field-name "group-id" :base-type :type/Integer :fk "group" :not-null? true}]
+    [["crowberto" 1]
+     ["rasta"     2]
+     ["lucky"     1]]]])
+
+(deftest action-error-handling-test
+  (mt/test-drivers (mt/normal-drivers-with-feature :actions)
+    (mt/dataset action-error-handling
+      (mt/with-actions-enabled
+        (mt/with-current-user (mt/user->id :crowberto)
+          (mt/with-actions [{_card-id :id}             {:dataset_query (mt/mbql-query group) :dataset true}
+                            {create-group :action-id} {:type :implicit
+                                                       :kind "row/create"}
+                            {update-group :action-id} {:type :implicit
+                                                       :kind "row/update"}
+                            {delete-group :action-id} {:type :implicit
+                                                       :kind "row/delete"}]
+            (mt/with-actions [{_card-id :id}           {:dataset_query (mt/mbql-query user) :dataset true}
+                              {create-user :action-id} {:type :implicit
+                                                        :kind "row/create"}
+                              {update-user :action-id} {:type :implicit
+                                                        :kind "row/update"}]
+              (testing "violate not-null constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking must have values."
+                          :errors {:ranking "You must provide a value."}}
+                         (mt/user-http-request :rasta :post 400 (format "action/%d/execute" create-group)
+                                               {:parameters {"name" "admin" "ranking" nil}}))))
+                (testing "when updating"
+                  (is (= "Implicit parameters must be provided."
+                         (mt/user-http-request :rasta :post 400 (format "action/%d/execute" update-group)
+                                               {:parameters {"id" "admin" "ranking" nil}})))))
+
+              (testing "violate unique constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking already exists."
```

Comment:
We should probably have a test for the plural form, when there are multiple columns involved.

Reply:
it's not easy to set up for that, we need to be able to define a not-null constraint for multiple columns, and it's currently not possible with our `defdataset`. 

it's good to cover that but I think it's too much effort.
I don't think we'll get that case a lot in the wild tho.

Reply:
I'm not sure it's that difficult to justify not covering it. We just need a table with composite PKs right? Here's an example.

https://github.com/metabase/metabase/blob/a47039c6b6d67aa811ae0d8e3f2c91195fe01e35/test/metabase/driver/mysql_test.clj#L448

You could copy+paste the code and change the json column to another type to make it easier.